x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/DomainConfig.java

```diff
@@ -0,0 +1,24 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+package org.elasticsearch.xpack.core.security.authc;
+
+import java.util.List;
+
+public record DomainConfig(String name, List<String> memberRealmNames, boolean literalUsername, String suffix)
+    implements
+        Comparable<DomainConfig> {
+
+    public DomainConfig {
+        assert (literalUsername && suffix != null) || (false == literalUsername && suffix == null);
+    }
+
+    @Override
+    public int compareTo(DomainConfig o) {
+        return name.compareTo(o.name);
+    }
+}
```

Comment:
Some minor, "nit" suggestions:
* Make the `List<String> memberRealmNames` a `Set<String>` (it bothers me that the `equals` method is not intuitive if we use the List)
* I don't think implementing `Comparable` is useful.
* I think it is possible to move the `literalUsername` and `suffix` validations inside the constructor.

Reply:
* I changed `memberRealmNames` to be a `Set`
* It is needed because the value of `realmToDomainsMap` is a [`TreeSet`](https://github.com/elastic/elasticsearch/blob/8c2f4292ec4b78e8c4eaec44443f0a32532cdf38/x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/security/authc/RealmSettings.java#L133). I inherited this code and it seems implementing `Comparable` is the least effort.
* I moved them into the constructor. I also tried to further beefy up `DomainConfig` to make it handle `Settings` in its constructor (similar to `RealmConfig`). But this causes cascading changes in tests and is a bit out of hand for this PR. Fortunately, it's all internal code. We can definitely refactor this later.